src/common/merkle_tree.zig

```diff
@@ -31,21 +31,17 @@ pub fn merkleTreeHash(hashes: []Hash, fanout: usize) !*Hash {
 pub const NestedHashTree = struct {
     hashes: []std.ArrayListUnmanaged(Hash),
 
-    pub fn getValue(self: *NestedHashTree, index: usize) !*Hash {
+    pub fn getValue(self: *NestedHashTree, index: usize) ?*Hash {
         var search_index: usize = 0;
-        var i: usize = 0;
-        while (i < self.hashes.len) {
-            const nested_len = self.hashes[i].items.len;
-            if (search_index + nested_len > index) {
+        for (self.hashes) |hash_list| {
+            if (search_index + hash_list.items.len > index) {
                 const index_in_nested = index - search_index;
-                return &self.hashes[i].items[index_in_nested];
+                return &hash_list.items[index_in_nested];
             } else {
-                search_index += nested_len;
-                i += 1;
+                search_index += hash_list.items.len;
             }
         }
-
-        return error.InvalidIndex;
+        return null;
```

Comment:
should still return an error - if you try to index into a list you get an out of bounds error, same logic should follow here

Reply:
If that's desired, the caller can write `tree.getValue(index) orelse return error.MoreSpecificErrorValue`.
If you index a slice/array out of bounds, you don't get an error, you get checked illegal behavior (panic in safe modes, UB in unsafe modes) - I would therefore argue it's orthogonal to the design of methods like this, where expressing the binary condition of "present or not present" simplifies its usage significantly. See the changes to `computeMerkleRoot` for example.

Reply:
feels like that responsibility shouldnt be for the caller to handle - imo it should behave the same as a slice/array so yeah ur right it should be illegal - maybe we should just panic

Reply:
Yeah I suppose we're not actually using the ability to recover from an invalid index anywhere.

Reply:
eb4981d7a3c1671ec26144ac7f153aeba3b85c1e